### PR TITLE
Add: _WD_ElementStyle

### PR DESCRIPTION
**Pull request by mlipok**
### Proposed changes

Adding support for set/get style property

### Checklist

Put an `x` in the boxes that apply. If you're unsure about any of them, don't hesitate to ask. We are here to help!<br>
This is simply a reminder of what we are going to look for before merging your code.

- [x] I have read and noticed the [CODE OF CONDUCT](https://github.com/Danp2/au3WebDriver/blob/master/docs/CODE_OF_CONDUCT.md) document
- [x] I have read and noticed the [CONTRIBUTING](https://github.com/Danp2/au3WebDriver/blob/master/docs/CONTRIBUTING.md) document
- [ ] I have added necessary documentation or screenshots (if appropriate)

### Types of changes

Please check `x` the type of change your PR introduces:

- [ ] Bugfix (change which fixes an issue)
- [x] Feature (change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (functional, structural)
- [ ] Documentation content changes
- [ ] Other (please describe)

### What is the current behavior?

there is no helper function to set/get style property

### What is the new behavior?

added `_WD_ElementStyle()`

### Additional context

https://www.autoitscript.com/forum/topic/205553-webdriver-udf-help-support-iii/?do=findComment&comment=1501634

### System under test

not related

### CODE REVIEW

**Comment by Danp2** [comment]
From [here](https://www.w3schools.com/tags/att_global_style.asp) --

image

so maybe use `$sCSSProperty` instead of `$sPropertyName` 🤔

**Comment by mlipok** [comment]
> > ;                  $sPropertyName - style property name which should be set, when is not defined (Default) then all properties are taken (array of properties will be returned)
> 
> You should consider alternatives for "property" here. I'm not sure if there's a standard name for these items within the style attribute.

[Parameters](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/getPropertyValue#parameters)
property is a [DOMString](https://developer.mozilla.org/en-US/docs/Web/API/DOMString) representing the property name (hyphen case) to be checked.

**Comment by mlipok** [comment]
> > ;                  $sValue - new value of the property to be set, when is not defined (Default) then properties are not get instead of set
> 
> Don't forget that this needs to support other data types beyond string.

in case when you want to set $sValue then it is only check if it is not `Default` keyword
```autoit
If IsString($sPropertyName) And $sValue <> Default Then ; set property value
```

btw.
Frome this documentation 
https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/getPropertyValue
<hr />
[Return value](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/getPropertyValue#return_value)
value is a [DOMString](https://developer.mozilla.org/en-US/docs/Web/API/DOMString) containing the value of the property. If not set, returns the empty string.
<hr />

Doesn't that mean that values are stored in the HTML DOM tree just as a string?

**Comment by Danp2** [comment]
> Doesn't that mean that values are stored in the HTML DOM tree just as a string?

IDK. I was thinking that you could have a style like this --

`style="display:none"`

I haven't tried it, but I thought your code would wrap the `none` in quotation marks.

**Comment by Danp2** [comment]
> For this reason I'm not sure about changing property to CSSProperty as I can directly copy getPropertyValue(property) from [style.getPropertyValue()](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/getPropertyValue) becuase this is common usage.

I think it works fine either way. I tend to prefer the later one since it better describes the parameter contents, but ultimately it's your decision. 🙂

**Review by Danp2 — COMMENTED** [review]
This isn't proper English. Suggested alternative --
 
 "new value to be set; current value will be retrieved when not supplied (Default)"

**Comment by mlipok** [comment]
> I haven't tried it, but I thought your code would wrap the `none` in quotation marks

Aha... I must check

**Comment by mlipok** [comment]
> I was thinking that you could have a style like this --
> 
> `style="display:none"`
> 
> I haven't tried it, but I thought your code would wrap the `none` in quotation marks.


If you are refering to this:
```autoit
_WD_ElementStyle($sSession, $sElement, 'display', 'none')
```
Then you must to know that this line
```autoit
$vResult = _WD_ExecuteScript($sSession, "var element = arguments[0]; element.style." & $sPropertyName & " = '" & $sValue & "'", __WD_JsonElement($sElement), Default, Default)
```

will be computed/evaluated to:
```autoit
$vResult = _WD_ExecuteScript($sSession, "var element = arguments[0]; element.style.display = 'none'", __WD_JsonElement($sElement), Default, Default)
```
so the JavaScript will be:
```javascript
element.style.display = 'none'
```

Is my reasoning correct?

**Comment by mlipok** [comment]
> ```js
> element.style.display = 'none'
> ```

I just check `.display = 'none'` on github website with this JS
```js
document.querySelector('ul.filter-list.js-notification-inboxes').style.display = 'none'
```

works well.

File: wd_helper.au3 [hunk 1]
```diff
@@ -1087,12 +1087,12 @@ EndFunc   ;==>_WD_ElementSelectAction
...
-; Syntax ........: _WD_ElementStyle($sSession, $sElement, $sPropertyName, $sValue)
+; Description ...: Set/Get element style CSSProperty
+; Syntax ........: _WD_ElementStyle($sSession, $sElement, $sCSSProperty, $sValue)
 ; Parameters ....: $sSession - Session ID from _WD_CreateSession
 ;                  $sElement - Element ID from _WD_FindElement
-;                  $sPropertyName - style/CSS property name which should be set. When it is undefined (Default) then all properties are taken (array of properties will be returned)
-;                  $sValue - new property value to be set. When it is undefined (default), properties are retrieved instead of setting them
+;                  $sCSSProperty - style property name which should be set, when is not defined (Default) then all properties are taken (array of properties will be returned)
+;                  $sValue - new value to be set; current value will be retrieved when not supplied (Default)
 ; Return values .: Success - Response from web driver in JSON format ...... #TODO
```
Comment (mlipok) [inline wd_helper.au3:1096]: Please propose changes in Return values (Success/Failure) descriptions

**Comment by mlipok** [comment]
> I was thinking that you could have a style like this --
> 
> `style="display:none"`
> 
> I haven't tried it, but I thought your code would wrap the `none` in quotation marks.

But on other hand.....
Wondering how to set such thing:

```css
font-family: 'Times New Roman', Times, serif
```

will try one of this:
```js
document.querySelector('ul.filter-list.js-notification-inboxes').style.font-family = "'Times New Roman', Times, serif"
document.querySelector('ul.filter-list.js-notification-inboxes').style.font-family = '"Times New Roman", Times, serif'
```

**Comment by mlipok** [comment]
here is proper usage:
```js
document.querySelector('ul.filter-list.js-notification-inboxes span.count').style.font-family == '"Lucida Console", "Courier New", monospace;'
```

so AutoIt code will be:
```autoit
_WD_ElementStyle($sSession, $sElement, 'font-family', '"Lucida Console", "Courier New", monospace;')
```

of course on https://github.com/notifications it will faill as you get:
```
Uncaught ReferenceError: family is not defined
    at <anonymous>:1:88
```

I think there is something that I do not understand in JavaScript that I must learn.

**Comment by mlipok** [comment]
Here are answers to my question:
https://stackoverflow.com/a/16601616/5314940

**Comment by mlipok** [comment]
I think we can add remark to `_WD_ElementStyle` about using `quotes` with parameter `$sValue`

**Comment by mlipok** [comment]
I work on some tests...

Using this
```js
GetPropertyValue();
function GetPropertyValue() {
	var elementtouse = '{"element-6066-11e4-a52e-4f735466cecf":"6ab37fa4-928b-4e76-b117-69f7b435805b"}';
	var search = 'display';
	var propertyname ='';
	for (let i = 0; i < elementtouse.style.length; i++) {
		propertyname = elementtouse.style.item(i);
		if (propertyname == search) {
			return elementtouse.style.getPropertyValue(propertyname);
			}
		}
	return '';
}
````

instead:
```autoit
Local Static $sJavaScript2 = _
		"var element = arguments[0];" & _
		"var search = '*';" & _
		"return GetPropertyValue(element, search);" & _
		"" & _
		"function GetPropertyValue(element, search) {" & _
		"	var propertyname ='';" & _
		"	for (let i = 0; i < element.style.length; i++) {" & _
		"		propertyname = element.style.item(i);" & _
		"		if (propertyname == search) {return element.style.getPropertyValue(propertyname);}" & _
		"	}" & _
		"	return '';" & _
		"}"
$vResult = _WD_ExecuteScript($sSession, StringReplace($sJavaScript2, '*', $sCSSProperty), __WD_JsonElement($sElement), Default, $_WD_JSON_Value)
```

I get this issues:

image

Could somebody tell me what I'm doing wrong ?

**Comment by Danp2** [comment]
Only the webdriver will understand`{"element-6066-11e4-a52e-4f735466cecf":"6ab37fa4-928b-4e76-b117-69f7b435805b"}`. You will need to use `querySelectorAll` or some other method to obtain the reference to the target element.

File: wd_helper.au3 (hunk 1 above)
Comment (mlipok) [reply wd_helper.au3:1096]: As for me this is the only place that should be reviewed.

**Comment by mlipok** [comment]
I already use this code and can say that is well tested, on both FF + Chrome.
Please make a finall review.

**Comment by Danp2** [comment]
> _WD_ElementStyle($sSession, $sElement, 'font-family', '"Lucida Console", "Courier New", monospace;')

This isn't working for me. Did you have a working version?

> I already use this code and can say that is well tested, on both FF + Chrome.

Can you share some additional code for testing?

**Comment by mlipok** [comment]
I try to change style on this page: `https://github.com/Danp2/au3WebDriver/pull/310`

I can set this using:
```js
document.querySelector('#repository-container-header').style.fontFamily = '"Lucida Console", "Courier New", monospace'
```

Try this code:

```autoit
Func UserTesting()
	_WD_Navigate($sSession, 'https://github.com/Danp2/au3WebDriver/pull/310')
	_WD_LoadWait($sSession)

	Local $sElement = _WD_FindElement($sSession, $_WD_LOCATOR_ByCSSSelector, '#repository-container-header')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information", 'Focuse on header (currently not changed)')

	_WD_ElementStyle($sSession, $sElement, 'fontFamily', '"Lucida Console", "Courier New", monospace')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	Local $sFontFamily = _WD_ElementStyle($sSession, $sElement, "font-family")
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information", $sFontFamily)

	Local $aListof_CSSProperties = _WD_ElementStyle($sSession, $sElement)
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	_ArrayDisplay($aListof_CSSProperties, '$aListof_CSSProperties')
;~ 	Exit
EndFunc   ;==>UserTesting
```

**Comment by mlipok** [comment]
I wonder why I get additional row in `$aListof_CSSProperties`
image

**Comment by mlipok** [comment]
should we add this following examples to function header ?
```autoit
; Example .......: _WD_ElementStyle($sSession, $sElement, 'fontFamily', '"Lucida Console", "Courier New", monospace')
; Example .......: _WD_ElementStyle($sSession, $sElement, "font-family")
; Example .......: _WD_ElementStyle($sSession, $sElement)
; Example .......: _WD_ElementStyle($sSession, $sElement, 'display', 'none')
```

File: wd_helper.au3 (hunk 1 above)
Comment (mlipok) [reply wd_helper.au3:1096]: For me allmost all here is fine, only this description.
Please @Danp2  propose some changes.

**Comment by Danp2** [comment]
> I wonder why I get additional row in $aListof_CSSProperties

You mean the blank row, correct?

**Comment by Danp2** [comment]
> should we add this following examples to function header ?

I would suggest limiting it to one example.

File: wd_helper.au3 (hunk 1 above)
Comment (Danp2) [reply wd_helper.au3]: See 5f2b2b6f2f4b3a4fde6d3a705a7613e0c1b9e704

Comment (mlipok) [reply wd_helper.au3]: Thanks for:
[Revise header](https://github.com/Danp2/au3WebDriver/pull/310/commits/5f2b2b6f2f4b3a4fde6d3a705a7613e0c1b9e704)

It looks good.

**Comment by mlipok** [comment]
> > I wonder why I get additional row in $aListof_CSSProperties
> 
> You mean the blank row, correct? You probably need to remove the trailing semicolon before calling _ArrayAdd.

I must only solve this and I hope this will be all ready to merge.

**Comment by mlipok** [comment]
new example:
```autoit
Func UserTesting()
	_WD_Navigate($sSession, 'https://github.com/Danp2/au3WebDriver/pull/310')
	_WD_LoadWait($sSession)

	; get element
	Local $sElement = _WD_FindElement($sSession, $_WD_LOCATOR_ByCSSSelector, '#repository-container-header')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information", 'Focus on header (currently not changed)')

	; add new fontFamily style
	_WD_ElementStyle($sSession, $sElement, 'fontFamily', '"Lucida Console", "Courier New", monospace')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	; check list of styles
	Local $aListof_CSSProperties_v1 = _WD_ElementStyle($sSession, $sElement)
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	_ArrayDisplay($aListof_CSSProperties_v1, '$aListof_CSSProperties_v1')

	; check single style
	Local $sFontFamily = _WD_ElementStyle($sSession, $sElement, "font-family")
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information", $sFontFamily)

	; remove 'fontFamily' style
	_WD_ElementStyle($sSession, $sElement, 'fontFamily', '')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	; remove 'backgroundColor' style
	_WD_ElementStyle($sSession, $sElement, 'backgroundColor', '')
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)

	; again - check list of styles
	Local $aListof_CSSProperties_v2 = _WD_ElementStyle($sSession, $sElement)
	ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & @CRLF)
	_ArrayDisplay($aListof_CSSProperties_v2, '$aListof_CSSProperties_v2')

	; additional check
	If Not UBound($aListof_CSSProperties_v2) Then MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information", 'All styles has been removed.')

EndFunc   ;==>UserTesting
```

**Comment by Danp2** [comment]
> I still wonder why once should be used `fontFamily` and in other cases `font-family` ?

I think it's a difference between JavaScript and CSS. See https://www.w3schools.com/cssref/pr_font_font-family.asp

**Comment by mlipok** [comment]
Thanks. 
So as for me this is ready to merge.

**Pull request merged**